### PR TITLE
#174848252 Create User Signup Endpoint

### PR DESCRIPTION
**Pull request by Adadie**
#### What does this PR do?
Have user sign up endpoint working

#### Description of Task to be completed?
- create userSignup route
- add JOI package
- add bcrypt package
- Hash password
- create signup validations
- create signup middleware

#### How should this be manually tested?

- Clone the repo git clone `https://github.com/atlp-rwanda/script-heroes-bn-backend.git`
- Switch your working directory by running `cd script-heroes-bn-backend`
- Checkout the branch `git checkout ft-user-registration-#174848252`
- Open the folder with you code editor
- Start your server
- Open postman
- Insert the localhost URL `http://localhost:3000/api/Signup`
- Send a post request using the signupModel provided below:
```
{
"firstName":"Enter name",
"lastName": "Enter last name",
"email": "Enter your email",
"phoneNumber": "Enter your phone number",
"password": "Enter Password"
}

```

#### What are the relevant pivotal tracker stories?
[#174797517](https://www.pivotaltracker.com/story/show/174797517)
[#174848252](https://www.pivotaltracker.com/story/show/174848252)

#### Screenshots (if appropriate)
Screen Shot 2020-09-24 at 14 43 44
Screen Shot 2020-09-24 at 14 44 08
Screen Shot 2020-09-24 at 14 44 33
Screen Shot 2020-09-24 at 14 47 16

### CODE REVIEW

File: src/controllers/signup.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import express from 'express';
+import bcrypt from 'bcryptjs';
+import signupvalidation from '../validations/signup';
+import { SignupModel } from '../database/models';
+
+const router = express.Router();
+
+// Register User
+
+router.post('/signup', async (req, res) => {
```
Comment (Habinezajanvier) [inline src/controllers/signup.js]: May it be good for us if you separate routes from controllers

File: src/database/migrations/20200924092914-create-signup-model.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+module.exports = {
```
Comment (Habinezajanvier) [inline src/database/migrations/20200924092914-create-signup-model.js]: Why are we having Users models and SignupModels, are you using both of it during signup

Comment (Adadie) [reply src/database/migrations/20200924092914-create-signup-model.js]: @Habinezajanvier One was created by @valens10 during testing. The one for signup is SignupModel

File: src/controllers/signup.js (hunk 1 above)
Comment (Adadie) [reply src/controllers/signup.js]: It's done

File: src/controllers/signup.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import bcrypt from 'bcryptjs';
+import signupvalidation from '../validations/signup';
+import { SignupModel } from '../database/models';
+import signupValidations from '../validations/signup';
+
+const router = express.Router();
+
+// Register User
+class signupController {
+  async signupUser(req, res) {
```
[the remark this answers is not in the document]
Comment (dushimeemma) [reply src/controllers/signup.js:10]: please fix this issues

File: src/controllers/signup.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import signupvalidation from '../validations/signup';
+import { SignupModel } from '../database/models';
+
+const router = express.Router();
+
+// Register User
+class signupController {
+  async signupUser(req, res) {
+    // Validate register before submitting
+    const { firstName, lastName, Email, phoneNumber, userPassword } = req.body;
```
[the remark this answers is not in the document]
Comment (dushimeemma) [reply src/controllers/signup.js:12]: please fix this issues

**Comment by BeyFaith** [comment]
Hi Adadie, great job..check your hound though. You have some violations.

**Comment by dushimeemma** [comment]
Try to send the token on register success @Junior-Adadie

**Review by BeyFaith — APPROVED** [review]

**Review by valens10 — COMMENTED** [review]
Great job.

**Comment by BeyFaith** [comment]
Adadie travis is failing

**Review by Habinezajanvier — APPROVED** [review]
ok

File: src/validations/signup.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      .min(8)
+      .regex(/[a-zA-Z]/)
+      .required()
+      .email()
+      .messages({
+        'string.base': 'Email name must be string',
+        'string.empty': 'Please fill in your email',
+        'string.min': 'email must be atleast {#limit} characters long',
+        'any.required': 'email is required'
+      }),
```
Comment (elvisiraguha) [inline src/validations/signup.js]: These error messages should use the correct property name, i.e if the missing property is _Email_ it shouldn't send _email_ is required.

File: README.md [hunk 6]
```diff
@@ -37,3 +37,31 @@ Note: If you make changes that uses environmental variables make sure to add tho
 ### 3 Running Seeds
 
 `sequelize db:seed:all`
+
+## Signup Endpoint
+
+### 1 Installing packages
+
+`npm i`
+
```
Comment (elvisiraguha) [inline README.md]: This is still unnecessary.

**Review by elvisiraguha — CHANGES_REQUESTED** [review]

**Review by dushimeemma — APPROVED** [review]

**Review by ericus123 — APPROVED** [review]

**Review by BeyFaith — APPROVED** [review]

**Review by elvisiraguha — APPROVED** [review]

**Comment by kipropbrian** [comment]
This works well, but there is a need for e-mail sending functionality.

**Pull request merged**